### ISSUE TITLE
Module request: Birddog studio

### ISSUE THREAD

**Issue by sytem** [bitfocus/companion-module-requests#103]
**Describe the feature**
Module for changing BD studio or Mini operating mode ( SDI/HDMI -> NDI or NDI -> SDI/HDMI) and active NDI input

**Is this platform dependent (windows, mac, ..)?**
no

**If documentation is required to implement, do you know where to find it?**

https://github.com/ualmarcus/BirdDog-cURL

**Usecases**
Using NDI to SDI convert as AUX for vmix or other NDI system

**Comment by McHauge** [comment]
Hey, personally I feel like I need just a tad more info before I can comfortably make a module. I'm a bit lost in the long commands/lists with no description, can I just phrase it as one long HTTP post request, or will I need to manipulate it? (normally I use TCP for most of the modules I make)

**Comment by josephdadams** [comment]
@McHauge take a look at my Teradek module. This looks like a similar approach. I have a birddog studio so I can help test.

**Comment by sytem** [comment]
Basicly there would be three actions. There need to be 3 variable for instance, ip/hostname, username and password

**Change NDI source**, where command is the first. ( I haven't yet tested those, just found them. Will do it next time I am at the studio I have those devices) As far I understand, each command is two http-requests, one to login and other for actual command.

`curl -s http://birddog-b334e.local/login -c cookiefile.txt -d "auth_password=birddog" && curl -b cookiefile.txt -s 'http://birddog-b334e.local/operation-configuration' -H 'Connection: keep-alive' -H 'Cache-Control: max-age=0' -H 'Content-Type: multipart/form-data; boundary=----WebKitFormBoundaryLrzvccuVYBrErN9Q' --data-binary $'------WebKitFormBoundaryLrzvccuVYBrErN9Q\r\nContent-Disposition: form-data; name="source_name"\r\n\r\n""\r\n------WebKitFormBoundaryLrzvccuVYBrErN9Q\r\nContent-Disposition: form-data; name="change_source_button"\r\n\r\nchange_source\r\n------WebKitFormBoundaryLrzvccuVYBrErN9Q--\r\n' --compressed
`

Parameter for this is the source name: "MSTHREESTICKM5 (Intel(R) HD Graphics 515 1)" These are global NDI source names, I don't know if they can be read from Birddog easily, or should there be just text input.


Other command is to **change to decode mode**, no need for parameters.

`curl -s http://birddog-b334e.local/login -c cookiefile.txt -d "auth_password=birddog" && curl -b cookiefile.txt -s 'http://birddog-b334e.local/operation-configuration' -H 'Connection: keep-alive' -H 'Cache-Control: max-age=0' -H 'Content-Type: multipart/form-data; boundary=----WebKitFormBoundaryA24wNG3dbt8rOKz7' --data-binary $'------WebKitFormBoundaryA24wNG3dbt8rOKz7\r\nContent-Disposition: form-data; name="mode"\r\n\r\nDecode\r\n------WebKitFormBoundaryA24wNG3dbt8rOKz7\r\nContent-Disposition: form-data; name="operation_mode_button"\r\n\r\noperation_mode\r\n------WebKitFormBoundaryA24wNG3dbt8rOKz7--\r\n' --compressed
`

And last one, **change unit to encode mode** from HDMI input

`curl -s http://birddog-b334e.local/login -c cookiefile.txt -d "auth_password=birddog" && curl -b cookiefile.txt -s 'http://birddog-b334e.local/video' -H 'Connection: keep-alive' -H 'Cache-Control: max-age=0' --data 'video-app-params=-H+AUTO+-a+-n+HDMI&video-form-path=video-param-grp-Manual-HDMI-Auto' --compressed && curl -b cookiefile.txt -s 'http://birddog-b334e.local/operation-configuration' -H 'Connection: keep-alive' -H 'Cache-Control: max-age=0' -H 'Content-Type: multipart/form-data; boundary=----WebKitFormBoundaryA24wNG3dbt8rOKz7' --data-binary $'------WebKitFormBoundaryA24wNG3dbt8rOKz7\r\nContent-Disposition: form-data; name="mode"\r\n\r\nEncode\r\n------WebKitFormBoundaryA24wNG3dbt8rOKz7\r\nContent-Disposition: form-data; name="operation_mode_button"\r\n\r\noperation_mode\r\n------WebKitFormBoundaryA24wNG3dbt8rOKz7--\r\n' --compressed
`

For this `'video-app-params=-H+AUTO+-a+-n+HDMI&video-form-path=video-param-grp-Manual-HDMI-Auto'` might be changing part when using Studio model ( https://www.bird-dog.tv/studio-overview/ ) , where there is both HDMI and SDI inputs

**Comment by josephdadams** [comment]
@sytem is the change mode to decode/encode still valid with the latest firmware? I updated my unit and it looks like it can do both at the same time.

**Comment by AlexK-99** [comment]
Hey, I found a backend interface on port 8080. You can call up the current source via the path 

> `"deviceIP:8080/connectTo"`

You can change the source with a post request on the same path. I have already programmed a test module for Companion and tested it with my birddog studio and birddog mini. I would finish the module and would request a repo to publish the module

**Comment by josephdadams** [comment]
@AlexK-99 hows this coming along?

**Comment by josephdadams** [comment]
@AlexK-99 we created this module repository for you but no code has ever been pushed to it. Are you interested in sharing your module? https://github.com/bitfocus/companion-module-birddog-studio

**Comment by AlexK-99** [comment]
@josephdadams Yes, I am interested in sharing my module. It's ready from my side, but the problem is that BirdDog hasn't officially released the new firmware required for it yet. I have received a beta version of the new firmware from Birddog, but it has bugs that make it unusable because you can't play the ndi sources.

**Comment by josephdadams** [comment]
What version?

**Comment by AlexK-99** [comment]
@josephdadams Version 3.0.8

**Comment by AlexK-99** [comment]
@josephdadams This version I got directly from BirdDog they are not yet official and are not stable and have a NDI bug.

**These versions have the full API, but as mentioned above, they have a NDI bug!!!**

Studio NDI:
 Firmware 3.0.8
[https://my.pcloud.com/publink/show?code=VZxpFnkZYebzL8uVI4p3WqSjfXKmY0iBlFzk](https://my.pcloud.com/publink/show?code=VZxpFnkZYebzL8uVI4p3WqSjfXKmY0iBlFzk)

Mini NDI:
Firmware 3.0.6
[https://my.pcloud.com/publink/show?code=XZSBhDkZIkmDHGfp1HhNMtyqt9bmRSKmI9KV](https://my.pcloud.com/publink/show?code=XZSBhDkZIkmDHGfp1HhNMtyqt9bmRSKmI9KV)

**Comment by AlexK-99** [comment]
I have uploaded the first version, which I think is finished. I have tested it with my BirdDog Studio / Mini.  It worked without problems. It should still be tested by someone else. I just added an action to change the Ndi decoding source. 
[https://github.com/bitfocus/companion-module-birddog-studio](https://github.com/bitfocus/companion-module-birddog-studio)

**Comment by AlexK-99** [comment]
BirdDog has finally released version **3.1**, I will test it and then finish the moudle, so you can change all settings

[https://www.bird-dog.tv/features/firmware-updates/](https://www.bird-dog.tv/features/firmware-updates/)
